File: tools/xlsxread/src/audit.rs
/// Audit subcommand: replicates audit-row-counts.js exactly.
///
/// Reads all .xlsx files from the input directory (sheet 0 only, matching the
/// JS script's behaviour at audit-row-counts.js:33), collects distinct SBDs
/// into a HashSet, then queries `SELECT COUNT(*) FROM student` from the DB.
/// Prints the same lines as audit-row-counts.js:54-62 and exits 0 on match,
/// 1 on mismatch.
use std::collections::HashSet;
use std::path::Path;

use calamine::{open_workbook_auto, Data, Reader};

use crate::config::DatasetConfig;
use crate::error::BuildError;
use crate::reader::is_header_row;

// ---------------------------------------------------------------------------
// Audit result
// ---------------------------------------------------------------------------

pub struct AuditResult {
    pub total_data_rows: u64,
    pub both_empty: u64,
    pub empty_name: u64,
    pub empty_sbd: u64,
    pub distinct_sbds: usize,
    pub db_count: i64,
    pub matched: bool,
}

// ---------------------------------------------------------------------------
// Main audit logic
// ---------------------------------------------------------------------------

/// Collect distinct SBDs from all xlsx files in `input_dir`, query `db_path`,
/// print the audit report and return the result.
///
/// The JS script reads only sheet 0 for every file (audit-row-counts.js:33).
/// Unlike build-database.js, the audit script does NOT iterate all sheets.
pub fn run_audit(
    input_dir: &Path,
    db_path: &Path,
    cfg: &DatasetConfig,
) -> Result<AuditResult, BuildError> {
    // Collect .xlsx files (audit-row-counts.js only checks .xlsx — line 15)
    let mut files: Vec<std::path::PathBuf> = std::fs::read_dir(input_dir)
        .map_err(|e| BuildError::Io {
            path: input_dir.display().to_string(),
            source: e,
        })?
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| {
            p.is_file()
                && p.extension()
                    .and_then(|e| e.to_str())
                    .map(|e| e.eq_ignore_ascii_case("xlsx"))
                    .unwrap_or(false)
        })
        .collect();
    files.sort();

    let mut all_sbd: HashSet<String> = HashSet::new();
    let mut total_data_rows: u64 = 0;
    let mut empty_name: u64 = 0;
    let mut empty_sbd: u64 = 0;
    let mut both_empty: u64 = 0;

    for file in &files {
        let path_str = file.display().to_string();
        let mut workbook = open_workbook_auto(file).map_err(|e| BuildError::Calamine {
            path: path_str.clone(),
            source: e,
        })?;

        let sheet_names = workbook.sheet_names().to_vec();
        if sheet_names.is_empty() {
            continue;
        }

        // audit-row-counts.js reads only sheet 0 (line 33: wb.SheetNames[0])
        let range =
            workbook
                .worksheet_range(&sheet_names[0])
                .map_err(|e| BuildError::Calamine {
                    path: path_str.clone(),
                    source: e,
                })?;

        let mut first_row = true;
        for raw in range.rows() {
            let row: Vec<Data> = raw.to_vec();

            // Skip header row on first row only
            if first_row {
                first_row = false;
                if is_header_row(&row, &cfg.header) {
                    continue;
                }
            }

            total_data_rows += 1;

            let ho_ten = row
                .get(cfg.columns.ho_ten)
                .map(|c| c.to_string().trim().to_owned())
                .unwrap_or_default();
            let sbd = row
                .get(cfg.columns.so_bao_danh)
                .map(|c| c.to_string().trim().to_owned())
                .unwrap_or_default();

            if ho_ten.is_empty() && sbd.is_empty() {
                both_empty += 1;
                continue;
            }
            if ho_ten.is_empty() {
                empty_name += 1;
            }
            if sbd.is_empty() {
                empty_sbd += 1;
            }
            if !sbd.is_empty() {
                all_sbd.insert(sbd);
            }
        }
    }

    // Query DB count
    let conn =
        rusqlite::Connection::open_with_flags(db_path, rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY)?;
    let db_count: i64 = conn.query_row("SELECT COUNT(*) FROM student", [], |row| row.get(0))?;

    let distinct_sbds = all_sbd.len();
    let matched = distinct_sbds as i64 == db_count;

    Ok(AuditResult {
        total_data_rows,
        both_empty,
        empty_name,
        empty_sbd,
        distinct_sbds,
        db_count,
        matched,
    })
}

// ---------------------------------------------------------------------------
// Print audit report — mirrors audit-row-counts.js:54-62 exactly
// ---------------------------------------------------------------------------

pub fn print_audit_report(r: &AuditResult) {
    println!("=== Source vs DB ===");
    println!(
        "Source: total data rows across all files: {}",
        r.total_data_rows
    );
    println!(
        "Source: rows with empty name AND sbd (skipped): {}",
        r.both_empty
    );
    println!("Source: rows with missing name only: {}", r.empty_name);
    println!("Source: rows with missing sbd only: {}", r.empty_sbd);
    println!("Source: distinct SBDs: {}", r.distinct_sbds);
    println!("DB:     row count: {}", r.db_count);
    println!(
        "Match:  {}",
        if r.matched {
            "YES — all unique SBDs accounted for".to_string()
        } else {
            format!("NO — gap of {}", r.distinct_sbds as i64 - r.db_count)
        }
    );
}

File: parser/internal/audit/audit.go
// Package audit compares distinct SBDs in the source spreadsheets against the
// row count in a built database.
//
// Two divergences from the build path are deliberate:
//
//   - only .xlsx files are considered
//   - only sheet 0 is read, regardless of sheet_mode
//
// They are intentional, not oversights. Do not "fix" them.
package audit

import (
	"database/sql"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/tiennm99/thptqg/parser/internal/config"
	"github.com/tiennm99/thptqg/parser/internal/ingest"
	"github.com/tiennm99/thptqg/parser/internal/reader"
	"github.com/tiennm99/thptqg/parser/internal/sqlitedb"
)

// Result carries the audit counters.
type Result struct {
	TotalDataRows uint64
	BothEmpty     uint64
	EmptyName     uint64
	EmptySbd      uint64
	DistinctSbds  int
	DBCount       int64
	Matched       bool
}

// Run collects distinct SBDs from the .xlsx files in inputDir and compares the
// total against the student row count in dbPath.
func Run(inputDir, dbPath string, cfg *config.DatasetConfig) (*Result, error) {
	entries, err := os.ReadDir(inputDir)
	if err != nil {
		return nil, fmt.Errorf("read input dir %s: %w", inputDir, err)
	}
	var files []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		if strings.EqualFold(filepath.Ext(e.Name()), ".xlsx") {
			files = append(files, filepath.Join(inputDir, e.Name()))
		}
	}
	sort.Strings(files)

	res := &Result{}
	seen := make(map[string]struct{})

	// Format-detection configs carry no Columns block, so the audit falls back
	// to fixed positional columns for them.
	hoTenCol, sbdCol := 1, 0
	if cfg.Columns != nil {
		hoTenCol, sbdCol = cfg.Columns.HoTen, cfg.Columns.SoBaoDanh
	}

	for _, file := range files {
		wb, err := reader.Open(file)
		if err != nil {
			return nil, err
		}
		sheets := wb.Sheets()
		if len(sheets) == 0 {
			wb.Close()
			continue
		}

		firstRow := true
		err = wb.EachRow(sheets[0].Index, func(_ reader.Sheet, _ int, row []reader.Cell) error {
			if firstRow {
				firstRow = false
				if ingest.IsHeaderRow(row, cfg.Header.Tokens) {
					return nil
				}
			}
			res.TotalDataRows++

			hoTen := cellAt(row, hoTenCol)
			sbd := cellAt(row, sbdCol)

			if hoTen == "" && sbd == "" {
				res.BothEmpty++
				return nil
			}
			if hoTen == "" {
				res.EmptyName++
			}
			if sbd == "" {
				res.EmptySbd++
			}
			if sbd != "" {
				seen[sbd] = struct{}{}
			}
			return nil
		})
		wb.Close()
		if err != nil {
			return nil, err
		}
	}

	// Opened read-only: the audit must never mutate the database it inspects.
	db, err := sql.Open(sqlitedb.DriverName, "file:"+dbPath+"?mode=ro")
	if err != nil {
		return nil, fmt.Errorf("open db %s: %w", dbPath, err)
	}
	defer db.Close()
	if err := db.QueryRow("SELECT COUNT(*) FROM student").Scan(&res.DBCount); err != nil {
		return nil, fmt.Errorf("count rows: %w", err)
	}

	res.DistinctSbds = len(seen)
	res.Matched = int64(res.DistinctSbds) == res.DBCount
	return res, nil
}

// PrintReport writes the source-versus-database comparison block.
func PrintReport(r *Result) {
	fmt.Println("=== Source vs DB ===")
	fmt.Printf("Source: total data rows across all files: %d\n", r.TotalDataRows)
	fmt.Printf("Source: rows with empty name AND sbd (skipped): %d\n", r.BothEmpty)
	fmt.Printf("Source: rows with missing name only: %d\n", r.EmptyName)
	fmt.Printf("Source: rows with missing sbd only: %d\n", r.EmptySbd)
	fmt.Printf("Source: distinct SBDs: %d\n", r.DistinctSbds)
	fmt.Printf("DB:     row count: %d\n", r.DBCount)
	if r.Matched {
		fmt.Println("Match:  YES — all unique SBDs accounted for")
	} else {
		fmt.Printf("Match:  NO — gap of %d\n", int64(r.DistinctSbds)-r.DBCount)
	}
}

func cellAt(row []reader.Cell, idx int) string {
	if idx < 0 || idx >= len(row) {
		return ""
	}
	return strings.TrimSpace(row[idx].Str)
}
